Improve memoizing of computed values

asterius1 opened the pull request:
This pull requests rewrites bunch of mobx core to 'change direction' of optimally recalculating changes.

Previously it was propagating from changed observables to observing derivations.

In proposed approach propagated up is only fact that some deep dependency might have changed (and reactions scheduled from it). Derivations before computing itself can check if any dependencies indeed changed and recompute when appropriate.
It introduces few big wins:
* When modifications and computations are mixed mobx can still rerun only needed stuff
* When in `autorun(() => a ? b : c)` `a` becomes true `c` is not being recomputed
* Batching of memoization can be introduced. For example `batch(() => { ... })` would make computed memoized for it's duration, giving (as far as I can tell) minimal possible number of recomputations regardless of modifications or reactions.

One issue I had is autoruns rerunning itself. Even though all tests for it are ok, I could point out some corner cases in which behaviour slightly change. In general self-rerunning autoruns, even outside this pull request, have many undefined/unexpected behaviours, although those cases can be quite artificial. I was thinking if we could add a rule so changes in autorun can't schedule it for rerun.

`computed temporary memoization` perf test previews the case, that made me do this in the first place.
Currently it takes 7 seconds. With changes it's down to ~1ms.
Most perf tests are very similar. Few perform worse.

The nicest thing about it is being able to not compromise clear code with performance even while implementing complex dependencies with possibly expensive actions.

Adjusted few tests, mainly lowering computation count.

Right now pull request doesn't pass the test for whyRun. Need to adjust language to new changes.

mweststrate commented:
So far this looks pretty amazing 😊. I will give detailed feedback and discuss stuff later. this might take a while but I'll come back on this ☺

mweststrate commented on src/core/computedvalue.ts:
```diff
@@ -20,18 +19,21 @@ export interface IComputedValue<T> {
+	newObserving = null; // during tracking it's array with new observed observers
+
+	isPendingUnobservation: boolean; // for effective unobserving
+	isObserved = false;
```
the following invariant if I'm correct? `isObserved === dependenciesState !== -1`? That might indicate that `isObserved` might not be explicitly needed?

asterius1 replied:
It is the case only for ComputedValue.
Atom doesn't implement derivations, so they don't have dependenciesState
Reaction doesn't implement observable so they don't have isObserved.

isObserved is there to make sure that clearing up batch doesn't flip out.
It's kinda safer with.
I am not sure if it's 100% necessary.

mweststrate commented on src/core/derivation.ts:
```diff
@@ -1,20 +1,25 @@
+	newObserving: IObservable[];
+	/**
+	 * Describes state of observing dependencies to know whet it's needed to rerun
+	 * -1 <- not tracking dependencies
```
using an enum would make this cleared: `enum IDerivationState { NOT_TRACKING = -1, UP_TO_DATE = 0, POSSIBLY_STALE = 1, STALE = 2 }`

mweststrate commented on src/core/derivation.ts:
```diff
@@ -24,6 +29,30 @@ export interface IDerivation extends IDepTreeNode, IObservable, ISetEntry {
+
+export function shouldCompute(derivation: IDerivation, onError?): boolean {
+	const dependenciesState = derivation.dependenciesState;
+	if (dependenciesState === 0) return false;
```
might benifit from a switch statement

mweststrate commented on src/core/derivation.ts:
```diff
@@ -24,6 +29,30 @@ export interface IDerivation extends IDepTreeNode, IObservable, ISetEntry {
+		for (let i = 0; i < derivation.observing.length; i++) {
+			const obj = derivation.observing[i];
+			if (obj instanceof ComputedValue) {
+				obj.get();
```
I see what this does, but calling `get()` and then relying on the side effects of `derivation.dependenciesState` might justify some elaborate comments :)

mweststrate commented on src/core/derivation.ts:
```diff
@@ -24,6 +29,30 @@ export interface IDerivation extends IDepTreeNode, IObservable, ISetEntry {
+			const obj = derivation.observing[i];
+			if (obj instanceof ComputedValue) {
+				obj.get();
+				if (derivation.dependenciesState === 2) return true;
```
the following invariant holds here, at this point the dependenciesState is either `2` or `0`?

asterius1 replied:
yep, gonna comment it.
`state === 1` is possible only on `ComputedValue`
at the end of `shouldCompute` `derivation.dependenciesState !== 1`
`ComputedValue.get` will always call shouldCompute

mweststrate commented on src/core/derivation.ts:
```diff
@@ -185,4 +197,58 @@ export function untrackedStart() {
+	const m = derivation.__mapid;
+
+	// just because it can be bottleneck
+	if (oldState === 0) {
```
this code could benefit from some de-dup. the suggested `move` seems eligable indeed.

asterius1 replied:
I'm right now trying to change whole `DerivationsSets` to something more performant, because it seems to be a performance bottleneck.

mweststrate commented on src/core/derivation.ts:
```diff
@@ -185,4 +197,58 @@ export function untrackedStart() {
+				delete o.observers.data2[m];
+				o.observers.data0[m] = derivation;
+			}
+		else // targetState === 1
```
does this ever happen?

mweststrate commented on src/core/derivation.ts:
```diff
@@ -136,37 +148,37 @@ function bindDependencies(derivation: IDerivation, prevObserving: IObservable[])
 	const l = obs.length;
 	for (let i = 0; i < l; i++)
 		removeObserver(obs[i], derivation);
+	derivation.dependenciesState = -1;
```
might be clearer to have changeDependencieState(-1) and clear the observers there?

mweststrate commented on src/core/derivation.ts:
```diff
@@ -185,4 +197,58 @@ export function untrackedStart() {
+				delete o.observers.data0[m];
+				o.observers.data1[m] = derivation;
+			}
+	}
```
conceptually could also transition to -1? should be handled here?

mweststrate commented on src/core/observable.ts:
```diff
@@ -1,10 +1,58 @@
 import {globalState} from "./globalstate";
-import {SimpleSet} from "../utils/set";
+
+export class DerivationsSets {
```
this obsoletes SimpleSet, so that one could be removed

mweststrate commented on src/core/observable.ts:
```diff
@@ -1,10 +1,58 @@
+
+export class DerivationsSets {
+	size: number = 0;
+	data0 = {};
```
I wouldn't be surprised, it is interesting to test that, that keeping one collection for all states and then filtering wherever the sets are used is faster? because `delete` is quite expensive and usually these sets are pretty constant over the lifetime of a derivation, so that would a lot of moving around.

mweststrate commented on src/core/observable.ts:
```diff
@@ -1,10 +1,58 @@
+			this.size--;
+		}
+	}
+	// move(targetState: number, value: IDerivation) {
```
+1 :) although if all is gathered in one set, the operation isn't needed at all. But not sure whether that is faster or terribly slow, let's investigate

mweststrate commented on src/core/globalstate.ts:
```diff
@@ -48,12 +48,9 @@ export class MobXGlobals {
-	 * After completing the transaction(s) these atoms will notify their observers.
-	 */
-	changedAtoms: IAtom[] = [];
+	inBatch: number = 0;
```
note that the version attribute needs to be bumped, as the globalState object's signature changes, meaning that this mobx version is not compatible with older mobx version in same memory

mweststrate commented on src/core/observable.ts:
```diff
@@ -16,46 +64,115 @@ export interface IObservable extends IDepTreeNode {
 
 export function addObserver(observable: IObservable, node: IDerivation) {
 	observable.observers.add(node);
+	if (!observable.isObserved) {
```
just `observable.isObserved =  true` ?

mweststrate commented on src/core/observable.ts:
```diff
@@ -16,46 +64,115 @@ export interface IObservable extends IDepTreeNode {
+			observable.isPendingUnobservation = false;
+			if (observable.isObserved && observable.observers.length === 0) {
+				observable.isObserved = false;
+				observable.onBecomeUnobserved(); // TODO: test if this happens only once, e.g. remove returns bool!
```
double check, this triggers clearObserving for all cases and results in state -1

asterius1 replied:
yep, it does for ComputedValues. Not for Atoms.

mweststrate commented on src/core/observable.ts:
```diff
@@ -16,46 +64,115 @@ export interface IObservable extends IDepTreeNode {
 	 * Check if last time this observable was accessed the same runId is used
 	 * if this is the case, the relation is already known
 	 */
+	observable.isObserved = true;
```
if i'm not mistaken, this can move inside the if. btw if that is the case, `addObserver` might not need to set `isObserved` as well?

mweststrate commented on src/core/reaction.ts:
```diff
@@ -112,6 +104,11 @@ export class Reaction implements IDerivation, IReactionPublic {
+	}
+
 	dispose() {
 		if (!this.isDisposed) {
```
reminder, add TODO on `whyRun` update

mweststrate commented on test/autorunAsync.js:
```diff
@@ -6,16 +6,21 @@ test('autorun 1', function(t) {
 		t.equal(_fired, fired, "autorun fired");
 		t.equal(_cCalcs, cCalcs, "'c' fired");
+		if(_cCalcs !== cCalcs) {
+			console.log('\n')
```
old debug statement?

mweststrate commented on test/observables.js:
```diff
@@ -1661,7 +1661,7 @@ test('unoptimizable subscriptions are diffed correctly', t => {
 	t.equal(b.observers.length, 1)
 	t.equal(c.observers.length, 1)
-	t.equal(d.$mobx.observing.length, 4) // 3 would be better!
+	t.equal(d.$mobx.observing.length, 3) // 3 would be better!
```
:+1:

mweststrate commented on test/observables.js:
```diff
@@ -1545,23 +1545,23 @@ test('forcefully tracked reaction should still yield valid results', function(t)
 
     t.equal(z, 4);
-    t.equal(runCount, 3);
+    t.equal(runCount, 2); // x is observed, so it should recompute only on dependency change
```
:+1

mweststrate commented:
Left a bunch of comments and notes to self for now, will look later a bit more in depth to a some individual tests and memory and cpu performance in general

asterius1 commented:
Thanks.
I am now rewriting DerivationsSets for better performance, because it's indeed a performance bottleneck.

asterius1 commented:
Made storing observers simpler and way more performant in common cases.
Also bumped globalState version and added some comments.

For performance it seems to be quite faster when data structure for observers is inlined into observable (without proxy set class). I added LegacyObservers, because some tests depended on it.
Probably should clear it up if possible. Is it likely that something depended on it?
Also can I just remove SimpleSet or some outside packagde might depend on it?

Or just depreciate?

mweststrate commented on src/core/computedvalue.ts:
```diff
@@ -32,8 +32,9 @@ export class ComputedValue<T> implements IObservable, IComputedValue<T>, IDeriva
+	lowestObserverState = 0;
 	unboundDepsCount = 0;
-	__mapid = "#" + getNextId();
+	__mapid = "#" + getNextId(); // NOT NEEDED ANYMORE
```
feel free to remove

mweststrate replied:
never mind :)

mweststrate commented on src/core/observable.ts:
```diff
@@ -65,6 +74,7 @@ export interface IObservable extends IDepTreeNode {
 	 */
 	lastAccessedBy: number;
 
+	lowestObserverState: number;
```
deserves some documentation :)

mweststrate commented on src/core/reaction.ts:
```diff
@@ -49,7 +51,9 @@ export class Reaction implements IDerivation, IReactionPublic {
 		if (!this._isScheduled) {
 			this._isScheduled = true;
 			globalState.pendingReactions.push(this);
+			startBatch();
```
push inside reactions?

mweststrate commented:
I might be using github wrong, but closing this PR in favor of #489 where further review comments can be processed.
